### ISSUE TITLE
Add Check Support for Java 21 Pattern Matching for Switch Syntax: New Check MissingNullCaseInSwitch

### ISSUE THREAD

**Issue by mahfouz72** [checkstyle/checkstyle#14979]
child of #14961 :

Switch statements and expressions in Java traditionally throw a `NullPointerException` if the selector expression evaluates to null. This necessitates an external null check before entering the switch. However, when Java introduces the new pattern matching for switch syntax, it becomes possible to integrate a `null` check within the switch statement by allowing a new null case label. This enhancement simplifies the code and eliminates the risk of `NullPointerException` by ensuring the null case is explicitly handled within the switch statement itself.

We should implement a new check named `MissingNullCaseInSwitch`. This check will enforce that any switch statement with a **reference type selector** must include a case null label.

We have a similar check ([EqualsAvoidNull](https://checkstyle.org/checks/coding/equalsavoidnull.html#EqualsAvoidNull)) to avoid a potential NullPointerException when using a string variable in LHS with `equals()`. We didn't make anything to allow suppression here even though it may be guaranteed that the LHS is not null.

---
**Config**
```XML
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="MissingNullCaseInSwitch"/>
    </module>
</module>
```
---
**Examples**

```Java
public class Test {
    public enum C {A}

    public void testPatternMatching(Point p) {
        switch (p) {  // violation
            case Point(int x, int y) -> System.out.println("error");
        }
        // after fix
        switch (p) {
            case Point(int x, int y) -> System.out.println("error");
            case null -> System.out.println("default");
        }
    }
    public void testStrings(String s) {
        switch (s) {  // violation
            case "a" -> System.out.println("a");
        }
    }
    
    public void testPrimitives(int i, char ch) {
        switch (i) {  // ok, NPE can't happen
            case 0 -> System.out.println("0");
        }
        switch (ch) {  // ok, NPE can't happen
            case 'a' -> System.out.println("a");
        }
    }
}

```
---

**Expected Output**
```powershell
$ java -jar checkstyle-10.19.0-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] D:\Test.java:5:9: Switch using reference types should have a null case. [MissingNullCaseInSwitch]
[ERROR] D:\Test.java:16:9: Switch using reference types should have a null case. [MissingNullCaseInSwitch]
[ERROR] D:\Test.java:22:9: Switch using reference types should have a null case. [MissingNullCaseInSwitch]
Audit done.
Checkstyle ends with 3 errors.
```

---
**Limitations**

switch can handle both primitive types and their corresponding boxed types. However, there is a  limitation when dealing with boxed types in switch. because we need to be aware of how to distinguishes between primitives and their boxed types.

**Example**  

```Java
	public void testBoxed(Integer i, char ch) {
		switch (i) {    
		        case 0 -> System.out.println("0");
                       case null -> System.out.println("1");
		}
	}
```
we can add case `null` in the above example because the switch selector is of a reference type (Integer). However, the check can't violate this example becuase we have no way to distinguitsh 0 (int) and their boxed type 0 (Integer)

---
 This check does not support enums; syntactically, these are no different from other constants.
**Example**  

```Java
     // Checkstyle is not type-aware. The
   // value of 'A' could be an enum constant, a primitive, or a reference type.
  void testEnums(E obj) {
        switch (obj) {  // Enum is a reference type it can be null
            case A -> {}
            case B -> {} 
            default -> {}
        }
    }
    public void visitToken(DetailAST ast) {
        switch (ast.getType()) { 
            case TokenTypes.METHOD_CALL:{} // this is just an integer can't be null
        }
    }
}
enum E {A, B, C}
class TokenTypes {
    public static final int METHOD_CALL = 2;
}
class DetailAST {
    public int getType() { return 0; }
}
```

**Comment by nrmancuso** [comment]
How can we know if we are dealing with a primitive or not, since we are not type aware?

**Comment by mahfouz72** [comment]
I think there is no need to be type-aware here. this can be known using AST only.

**Record patterns**
```
        |       |   |--SWITCH_RULE -> SWITCH_RULE [6:12]
        |       |   |   |--LITERAL_CASE -> case [6:12]
        |       |   |   |   `--RECORD_PATTERN_DEF -> RECORD_PATTERN_DEF [6:17]
        |       |   |   |       |--MODIFIERS -> MODIFIERS [6:17]
```
---
**Enums**
```
        |       |   |--SWITCH_RULE -> SWITCH_RULE [12:12]
        |       |   |   |--LITERAL_CASE -> case [12:12]
        |       |   |   |   `--EXPR -> EXPR [12:17]
        |       |   |   |       `--IDENT -> A [12:17]
```
---
**Strings**
```
        |       |   |--SWITCH_RULE -> SWITCH_RULE [18:12]
        |       |   |   |--LITERAL_CASE -> case [18:12]
        |       |   |   |   `--EXPR -> EXPR [18:17]
        |       |   |   |       `--STRING_LITERAL -> "a" [18:17]
```
---
**Primitives** (int, char)
```
        |       |   |   |--LITERAL_CASE -> case [24:12]
        |       |   |   |   `--EXPR -> EXPR [24:17]
        |       |   |   |       `--NUM_INT -> 0 [24:17]

```
```
        |       |   |--SWITCH_RULE -> SWITCH_RULE [27:12]
        |       |   |   |--LITERAL_CASE -> case [27:12]
        |       |   |   |   `--EXPR -> EXPR [27:17]
        |       |   |   |       `--CHAR_LITERAL -> 'a' [27:17]

```

**Comment by nrmancuso** [comment]
@mahfouz72 good thinking.

I think the message needs some work, though. What about “Switches using reference types should have a null case”?

Another item for discussion: should we allow suppression of this check by annotation on method parameter?

Another: what about methods that guarantee a non null return value? I cannot think of any way to suppress this without in-code suppressions.

**Comment by mahfouz72** [comment]
> I think the message needs some work, though. What about “Switches using reference types should have a null case”?

yes, this message is better but can it be only "Switch" without 's' I think the plural is not necessary here.



> Another item for discussion: should we allow suppression of this check by annotation on method parameter?

do you mean something like [SuppressWarningsFilter](https://checkstyle.org/filters/suppresswarningsfilter.html)  ?


I would say what did we do on similar checks in the past to handle these cases when enforcing something is guaranteed to be not  necessary.

I found [EqualsAvoidNull](https://checkstyle.org/checks/coding/equalsavoidnull.html#EqualsAvoidNull) we didn't make anything to allow suppression here even though it may be guaranteed that the LHS is not null. 

> I cannot think of any way to suppress this without in-code suppressions.

can't it be suppressed using XpathFilters?

**Comment by nrmancuso** [comment]
>yes, this message is better but can it be only "Switch" without 's' I think the plural is not necessary here.

I used "switches"  to take the place of "switch expressions and statements"

>do you mean something like [SuppressWarningsFilter](https://checkstyle.org/filters/suppresswarningsfilter.html) ?

Example:

```java
void m(@NotNull Integer i) {
  switch (i) { // is ok since method parameter is annotated with @NotNull
     case Integer x when x > 0: ...
  ...
  }
}
```

We can think about this for second iteration of this check, but it is something to keep in mind for sure.

>I found [EqualsAvoidNull](https://checkstyle.org/checks/coding/equalsavoidnull.html#EqualsAvoidNull) we didn't make anything to allow suppression here even though it may be guaranteed that the LHS is not null.

Precedence by existing checks is good, please add this to the issue description

> xpath filters

This is not a good solution for a check like this, we would probably want some in-code suppressions for this since it would only be certain cases.

**Comment by nrmancuso** [comment]
I am good to approve for the most basic functionality with the name `MissingNullCaseInSwitch` and message `Switch using reference types should have a null case.`

@rnveach please place approved label if you agree

**Comment by rnveach** [comment]
>             case Point(int x, int y) -> System.out.println("error");

This doesn't compile on Eclipse.

> How can we know if we are dealing with a primitive or not, since we are not type aware?
>     public void testPrimitives(int i, char ch) {
        switch (i) {  // ok, NPE can't happen

What about the following? I was able to compile this in Eclipse.
````
	public void testClassPrimitives(Integer i, char ch) {
		switch (i) {
		case 0 -> System.out.println("0");
		}
		switch (i) {
		case 0 -> System.out.println("0");
		case null -> System.out.println("1");
		}
````
Notice `i` is not an primitive but the case is a primitive. It matches your primitive switch.

**Comment by nrmancuso** [comment]
>Notice i is not an primitive but the case is a primitive. It matches your primitive switch and I can add a null case.

Good find, @rnveach. I think this false negative is ok, as long as we call out this limitation.

**Comment by mahfouz72** [comment]
hmm, good catch I think for this example we have to be type-aware. 
we can't  distinguish between 0 (int) and 0 (Integer) , c (char) and c (Character)

**Comment by rnveach** [comment]
> this false negative is ok, as long as we call out this limitation
> We have a similar check ([EqualsAvoidNull](https://checkstyle.org/checks/coding/equalsavoidnull.html#EqualsAvoidNull)) to avoid a potential NullPointerException when using a string variable in LHS with equals().

I wanted to point out that [EqualsAvoidNull also has a limitation that it must be able to identify what it is looking at is a String, otherwise it is ignored](https://github.com/checkstyle/checkstyle/blob/512895e18c5f07b18311e0c317cdea5dc13f4969/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/EqualsAvoidNullCheck.java#L325).

So I am good with this limitation and to approve this issue.

@mahfouz72 Please update first post to add documentation on this limitation.

> we can't distinguish between 0 (int) and 0 (Integer) , c (char) and c (Character)

All primitives have a class version. So there is also float, long, boolean, void, etc...

**Comment by mahfouz72** [comment]
> @mahfouz72 Please update first post to add documentation on this limitation.

done

> All primitives have a class version. So there is also float, long, boolean, void, etc...

yes and switch selectors can be int or char only

**Comment by rnveach** [comment]
> yes and switch selectors can be int or char only

Is the selector the thing be switched on or is it the case?

This compiles in Eclipse:
````
	public void test(float i) {
		switch (i) { // ok, NPE can't happen
		case 0.0f -> System.out.println("0");
		default -> {}
		}
````

**Comment by mahfouz72** [comment]
```
public class Test {
    public void testPrimitives(float f) {
        switch (f) {
            case 0.0f -> System.out.println("0");
       }
    }
}
PS D:\CS\test> javac src/Test.java
src\Test.java:3: error: selector type float is not allowed
        switch (f) {
               ^
1 error
PS D:\CS\test> 

```
what is your Java version? I thought that float was never supported in switch

**Comment by rnveach** [comment]
I just downloaded `jdk-22.0.1` today to do this issue. Please note, if my memory serves me, Eclipse uses its own compiler.

I am using Eclipse 2023-06, which seems to let me only set project levels to Java 20 at the most. I am waiting on an external source to upgrade to 2024-06 before I update too, so it could be a bug already fixed.

**Comment by mahfouz72** [comment]
@nrmancuso @rnveach 

while starting to think more about this check (and already make the basic implementation) I caught another limitation

```
    void testEnums(E obj) {
        switch (obj) {  // Enum is a reference type it can be null
            case E.A -> {}
            case B -> {}
            default -> {}
        }
    }
    public void visitToken(DetailAST ast) {
        switch (ast.getType()) { 
            case TokenTypes.METHOD_CALL:{} // this is just an integer can't be null
        }
    }
}
enum E {A, B, C}
class TokenTypes {
    public static final int METHOD_CALL = 2;
}
class DetailAST {
    public int getType() { return 0; }
}
```

for this example we must be type-aware to differeniate between both switchs (the ast is the same). So if we implement a check to violate the code of the first method the second code will produce false positive

Are we ok with this false positive and call it out ? or don't violate the case labels that is IDENT at all and at this case we will have false negative in enum constant in case label

**Comment by nrmancuso** [comment]
@mahfouz72 @rnveach let's do the most simple case only for the initial implementation: switch rules/case groups with pattern definitions only.

**Comment by mahfouz72** [comment]
> let's do the most simple case only for the initial implementation: switch rules/case groups with pattern definitions only.

Agreed + String Literal

**Comment by rnveach** [comment]
> let's do the most simple case only for the initial implementation: switch rules/case groups with pattern definitions only.

If there are other cases we wish to add support to, we should move them to a new issue so they are not forgotten and explain what is missing. It is probably better to do this as we work on this before merge so it isn't forgotten on a merge.
